README.md
# [jquery-pjaxr](https://www.minddust.com/project/jquery-pjaxr/) [![Bower version](https://badge.fury.io/bo/jquery-pjaxr.png)](http://badge.fury.io/bo/jquery-pjaxr) [![Build Status](https://secure.travis-ci.org/minddust/jquery-pjaxr.png)](http://travis-ci.org/minddust/jquery-pjaxr) [![devDependency Status](https://david-dm.org/minddust/jquery-pjaxr/dev-status.png)](https://david-dm.org/minddust/jquery-pjaxr#info=devDependencies)

[![Selenium Test Status](https://saucelabs.com/browser-matrix/minddust.svg)](https://saucelabs.com/u/minddust)

jquery-pjaxr is [jQuery](http://jquery.com/) plugin that uses ajax and pushState to deliver a fast browsing experience.

It is able to replace **multiple containers** and different **head tags** with full back-forward functionality.

For [browsers that don't support pushState](http://caniuse.com/#search=pushstate) pjaxr fully degrades.

**Finally released version 1.2 but there is still a lot of work to do and currently no time for me to maintain this lib.**


## Why another pjax library?

Shortly after starting with pjax a ran into some limitations like:

* how to add additional head elements?
* how to update multiple containers?

I started writing a workaround which became an own library and here we are :).

I hope pjaxr will help people who are running into the same problems like I did.


## Note

There is already an awesome plugin called [jquery-pjax](https://github.com/defunkt/jquery-pjax) on which this project is based.  For not breaking existing frontends but to support the same port to the server this library is named **pjaxr** and accepts the same header **X-PJAX**.

> There are still many testcases missing but the most signicant are setup. PR's are welcome.


## Installation

* Download the latest release: [v1.2.0](https://github.com/minddust/jquery-pjaxr/archive/v1.2.0.zip)
* Clone the repository: `git clone git@github.com:minddust/jquery-pjaxr.git`.
* Curl the library: `curl -O https://raw.github.com/minddust/jquery-pjaxr/master/jquery.pjaxr.min.js`
* Install with [Bower](http://bower.io): `bower install jquery-pjaxr`.


## Dependencies

Requires jQuery 1.9.x or higher.


## Usage

```javascript
$(document).pjaxr('a');
```

That's all you need to activate pjaxr functionality.

If you only want to bind pjaxr when it's supported by the user's browser, you can activate pjaxr like this:

```javascript
if ($.support.pjaxr) {
    $(document).pjaxr('a');
}
```

Or you can call the click handler by yourself and wrap it with some additional start functionality like this:

```javascript
if ($.support.pjaxr) {
    $('#some-menu').on('click', 'a[data-pjaxr]', function(event) {
        $(document).pjaxr.click(event);
    });
}
```

Or you can make a pjaxr request by pure calling:

```javascript
$(document).pjaxr.request('/about/', {timeout: 1337});
```

If you are migrating an existing site you probably don't want to enable pjaxr everywhere just yet.  Instead of using a global selector like `a` try annotating pjaxrable links with `data-pjaxr`, then use `'a[data-pjaxr]'` as your selector.

### data-pjaxr-namespace

There is also the possibility to add a `data-pjaxr-namespace` to the body which will then be set as `X-PJAX-NAMESPACE` in the request header.  This is just for initialising the namespace - Any further values should come from pjaxr responses via `<pjaxr-namespace>`.

By passing the namespace it's possible to limit the replacement area on the server side.  Checkout existing pjaxr libraries below.


## Settings

Of course there are some options which will change your pjaxr behavior:

* `timeout`: the time in ms pjaxr will wait for a server response before hard loading the page. (default: `650`)
* `push`: determines whether to push the pjaxr request or not (default: `true`)
* `replace`: determines whether to replace the history state or not. will be ignored if `push` is `true`. (default: `false`)
* `scrollTo`: position in pixel the to scroll after pjaxring. (default: `0`)
* `version`: delivered pjaxr version. used to compare with `X-PJAX-VERSION` of the response header to force hard load on mismatch. (default: `<meta http-equiv="X-PJAX-VERSION" content="...">`)

You can either pass them as a second parameter on your `pjaxr` call or override them globally via `$.fn.pjaxr.defaults`.


## Signals

jquery-pjaxr fires a number of events regardless of how its invoked.

All events are fired from the document, cause the actions concern the whole page.

### start and end

* `pjaxr:start` - Fired when a pjaxr request is made or `popstate` is triggered.
* `pjaxr:end` - Fired when a pjaxr request ends or `popstate` is triggered.
* `pjaxr:click` - Fired when a pjaxr link is clicked.

### ajax related

* `pjaxr:beforeSend` - Fired before the pjaxr request begins. Preventing this event will abort the request.
* `pjaxr:send` - Fired after the pjaxr request begins.
* `pjaxr:always` - Fired after the pjaxr request finishes.
* `pjaxr:success` - Fired after the pjaxr request succeeds.
* `pjaxr:done` - Fired after the pjaxr request is processed successfully.
* `pjaxr:fail` - Fired after the pjaxr request fails. Preventing this event will disable fallback redirect.
* `pjaxr:timeout` - Fired if after timeout is reached. Preventing this event will disable the fallback and will wait indefinitely until the response returns.

`send` and `complete` are a good pair of events to use if you are implementing a loading indicator. They'll only be triggered if an actual request is made, not if it's loaded from cache.

```javascript
$(document).on('pjaxr:send', function() {
  $('#loading').show()
})
$(document).on('pjaxr:complete', function() {
  $('#loading').hide()
})
```

## Response Structure & Rules

Check if the request header have **X-PAX** set and return rendered html in like this format:

```html
<pjaxr-head>
    <title>...</title>
    <meta name="..." ...>
    <meta property="..." ...>
    <link href="..." ...>
    <script src="..."></script>
    <style>...</style>
</pjaxr-head>
<pjaxr-body>
    <... id="..." ...></...>
    <... id="..." ...></...>
    <... id="..." ...></...>
</pjaxr-body>
<pjaxr-namespace>...</pjaxr-namespace>
```

* `pjaxr-head` and `pjaxr-body` are optional but if both are missing, pjaxr will hard load the giving url.
* `pjaxr-head`
    * `title` will always be replaced if given.
    * `meta` will be replaced if `name` or `property` find a match, otherwise it will be appended.
    * `link` will be replaced if `href` finds a match, otherwise it will be appended.
    * `script` will be replaced if `href` finds a match, otherwise it will be appended.
    * `style` will always be appended.
    * `data-remove-on-pjaxr` can the written to any tag and will force the element to be removed with the next pjaxr request and restored on popstate.
* `pjaxr-body`
    * every child must have an id.
* `pjaxr-namespace` is optional but when given - it's value will be passed on further request via `X-PJAX-NAMESPACE`.
* any other content will be ignored.


### Existing Pjaxr Server Side Libraries

These libs are making use of the advanced pjaxr namespacing functionality:

* Django: https://github.com/iekadou/django-pjaxr


### Existing Pjax Server Side Libraries

There are many available plugins for different languages and frameworks which will lift the heavy work for you:

* Asp.Net MVC3: http://biasecurities.com/blog/2011/using-pjax-with-asp-net-mvc3/
* Aspen: https://gist.github.com/whit537/6059536
* CakePHP : https://github.com/sanojimaru/CakePjax
* Django: https://github.com/jacobian/django-pjax
* Express: https://github.com/abdelsaid/express-pjax-demo
* Flask: https://github.com/zachwill/pjax_flask
* FuelPHP: https://github.com/rcrowe/fuel-pjax
* Grails: http://www.bobbywarner.com/2012/04/23/add-some-pjax-to-grails/
* Rails: https://github.com/rails/pjax_rails


## Testing

As noted above: the testcases are **incomplete**

The main reason for this is that it's really hard to simulate complex page transitions. If you want to support this project - you can take a look at the structure of the existing ones an write some additional tests. Help is always welcome!

To run the tests locally you have to follow these steps:

1. clone the repository

    ```
git clone git@github.com:minddust/jquery-pjaxr.git
    ```
2. install python dependencies for test suite

    ```
pip install Django==1.6.6
pip install django-pjaxr==1.1.1
pip install selenium==2.42.1
    ```
3. install node dependencies for test suite

    ```
npm install
    ```
4. copy pjaxr js into test_app

    ```
npm run-script prepare-tests
    ```
5. install node dependencies for test_app (bower)

    ```
cd test_app
npm install
    ```
6. install bower dependencies for test_app

    ```
npm run-script bower-install
    ```
7. run the tests

    ```
cd ..
python test_app/tests/runtests.py
    ```
    
Note: you may need to download the [ChromeDriver](http://chromedriver.storage.googleapis.com/index.html) if it's not already installed. 

## Sites using jquery-pjaxr

If you are using this library and want to get listed below.  Please let me know.  Just make a pull request or write me an [email](http://www.google.com/recaptcha/mailhide/d?k=011zym_c1aKrhZC1K3TKFvVQ==&c=fPxx09KG4QgsFYu7CWApi-ZBAupHsLalyCVJ_Vogc8g= "Reveal my email address").

* https://www.socialfunders.org
* http://www.noxic-action.de/page/


## Contributing

Help is appreciated!


## Thanks

I like to thank [Chris Wanstrath](https://github.com/defunkt) for his really awesome [jquery-pjax](https://github.com/defunkt/jquery-pjax) library.  This project wouldn't exist without his work.


## Copyright and license

Copyright 2013-2015 Stephan Groß, under [MIT license](https://github.com/minddust/jquery-pjaxr/blob/master/LICENSE).

Want to appreciate my work? [minddust at Gittip](https://www.gittip.com/minddust/)


CHANGELOG.md
# Master

* Add saucelabs integration
* Add pjaxrReady and pjaxrAlways register methods #29
* Update jQuery dependency to >= 1.9 #31
* Fix #12 ignore unprocessable metatags
* Fix #16 chrome/webkit caching bug
* Fix #17 misuse of location.replace
* Fix #19 missing pjaxr:send event
* Fix #24 strip namespace processing
* Fix #38 missing testing instructions

# 1.1.0 (2013-11-21)

* Fix #10 cross origin checks for low IE versions

# 1.0.6 (2013-08-31)

* Fix #8 by adding try-catch block in body processing

# 1.0.5 (2013-08-27)

* Add pure request calls
* Remove options from pjaxr:click event
* Fix namespace replacement on popstate

# 1.0.4 (2013-08-22)

* Add contributors list
* Fix #6 missing related target

# 1.0.3 (2013-08-22)

* Add `success` signal
* Add `pjaxr-namespace` handling
* Fix total brick in 1.0.2 (wrong jQuery selector caused script to stop)

# 1.0.2 (2013-08-21)

* Add support to pass pjaxr namespace

# 1.0.1 (2013-08-19)

* Add support to call click handling directly
* Fix head tag handling typo

# 1.0.0 (2013-08-08)

* Fix minor bugs

# 1.0.0-beta.1 (2013-08-03)

* Birth of jquery-pjaxr


tests/app.py
from __future__ import unicode_literals

import os

from flask import Flask, render_template, send_from_directory, request


PROJECT_ROOT = os.path.abspath(os.path.dirname(__file__))


app = Flask(__name__)


# thanks @zachwill - https://github.com/zachwill/pjax_flask/blob/master/app.py
def pjax(template):
    if 'X-PJAX' in request.headers:
        return render_template(_pjaxify_template_var(template))
    return render_template(template)


# thanks @jacobian - https://github.com/jacobian/django-pjax/blob/master/djpjax.py
def _pjaxify_template_var(template_var):
    if isinstance(template_var, (list, tuple)):
        template_var = type(template_var)(_pjaxify_template_name(name) for name in template_var)
    elif isinstance(template_var, basestring):
        template_var = _pjaxify_template_name(template_var)
    return template_var


def _pjaxify_template_name(name):
    if '.' in name:
        name = '%s-pjax.%s' % tuple(name.rsplit('.', 1))
    else:
        name += '-pjax'
    return name


@app.route('/')
def md_qunit():
    return render_template('qunit.html')


@app.route('/favicon.ico')
def md_favicon():
    return send_from_directory(PROJECT_ROOT, 'favicon.ico')


@app.route('/jquery.pjaxr.js')
def md_pjaxr():
    return send_from_directory(os.path.join(PROJECT_ROOT, '..'), 'jquery.pjaxr.js')


@app.route('/eval.js')
def md_eval():
    return send_from_directory(PROJECT_ROOT, 'eval.js')


@app.route('/<path:filename>/')
def md_template(filename):
    return pjax('{filename}.html'.format(filename=filename))


@app.route('/libs/<path:filename>')
def md_lib(filename):
    return send_from_directory(os.path.join(PROJECT_ROOT, 'libs'), filename)


@app.route('/qunit/<path:filename>')
def md_test(filename):
    return send_from_directory(os.path.join(PROJECT_ROOT, 'qunit'), filename)


if __name__ == '__main__':
    app.run()


tests/eval.js
window.evaledSrcScript = true;

tests/qunit/test.js
if ($.support.pjaxr) {
    module('$.pjaxr', {
        setup: function() {
            var self = this;
            stop();
            window.iframeLoad = function(frame) {
                self.frame = frame;
                self.iframe = $('iframe').contents();
                window.iframeLoad = $.noop;
                start();
            };
            $('#qunit-fixture').append('<iframe src="/home/">');
        },
        teardown: function() {
            delete window.iframeLoad;
        }
    });

    asyncTest('pushes new url', 1, function() {
        var iframe = this.iframe;

        $(document).on('pjaxr:end', function() {
            equal(window.location.pathname, '/blog/');

            start();
        });

        $('body', iframe).pjaxr('a[data-pjaxr]');
        $('a[href="/blog/"]', iframe).trigger("click");
    });
}

Gulpfile.js
var gulp = require('gulp');
var header = require('gulp-header');
var rename = require('gulp-rename');
var rimraf = require('gulp-rimraf');
var uglify = require('gulp-uglify');
var util = require('gulp-util');

var pkg = require('./package.json');
var currentYear = util.date(new Date(), 'yyyy');

var banner = '/*! <%= pkg.name %> v<%= pkg.version %> | Copyright (c) 2013-<%= currentYear %> <%= pkg.author %> | <%= pkg.license %> license | <%= pkg.homepage %> */\n';

gulp.task('scripts', function() {
    return gulp.src('./jquery.pjaxr.js')
        .pipe(uglify())
        .pipe(header(banner, {pkg: pkg, currentYear: currentYear}))
        .pipe(rename({suffix: '.min'}))
        .pipe(gulp.dest('.'));
});

gulp.task('clean', function() {
    return gulp.src('./jquery.pjaxr.min.js', {read: false})
        .pipe(rimraf({force: true}));
});

gulp.task('prepare-tests', function() {
    return gulp.src('./jquery.pjaxr.js')
        .pipe(gulp.dest('./test_app/static/'))
});

gulp.task('default', ['clean', 'scripts']);


test_app/urls.py
from __future__ import unicode_literals

from django.conf.urls import patterns, url
from django.views.generic import TemplateView, RedirectView


urlpatterns = patterns('',
    url(r'^$', TemplateView.as_view(template_name='index.html'), name='index'),
    url(r'^about/$', TemplateView.as_view(template_name='about.html'), name='about'),
    url(r'^ignored-metatag/$', TemplateView.as_view(template_name='ignored_metatag.html'), name='ignored_metatag'),
    url(r'^no-matching-id/$', TemplateView.as_view(template_name='no_matching_id.html'), name='no_matching_id'),
    url(r'^no-pjaxr-response/$', TemplateView.as_view(template_name='no_pjaxr_response.html'), name='no_pjaxr_response'),
    url(r'^pjaxr-ready-pjaxr-always/(?P<pjaxr_state>(disabled))?/$', TemplateView.as_view(template_name='pjaxr_ready_pjaxr_always.html'), name='pjaxr_ready_pjaxr_always'),
    url(r'^project/$', TemplateView.as_view(template_name='project.html'), name='project'),
    url(r'^project/blog/$', TemplateView.as_view(template_name='project_blog.html'), name='project_blog'),
    url(r'^project/gallery/$', TemplateView.as_view(template_name='project_gallery.html'), name='project_gallery'),
    url(r'^settings/$', TemplateView.as_view(template_name='settings.html'), name='settings'),

    # prevent travis 404 logs
    url(r'^favicon\.ico$', RedirectView.as_view(url='/static/favicon.ico')),
)


jquery.pjaxr.js
/*!
 * jquery.pjaxr v1.2.0 by @minddust
 * Copyright (c) 2013-2014 Stephan Groß
 *
 * http://www.minddust.com/project/jquery-pjaxr
 *
 * Licensed under the MIT license:
 * http://www.opensource.org/licenses/MIT
 */
(function ($) {
    'use strict';

    function fnPjaxr(selector, options) {
        return this.on('click.pjaxr', selector, function (event) {
            handleClick(event, options);
        });
    }

    function fnPjaxrReady(func) {
        // if document is not ready yet --> the request must be initial --> register on $(document).ready()
        if (!$.isReady) {
            return $(document).ready(func);
        }
        // else --> the request must come from pjaxr --> register on $(document).one('pjaxr:done');
        else {
            return $(document).one('pjaxr:done', func);
        }
    }

    function fnPjaxrAlways(func) {
        // same as fnPjaxrReady but will always be called after a request is processed
        if (!$.isReady) {
            $(document).ready(func);
        }
        return $(document).on('pjaxr:done', func);
    }

    function request(link, options) {
        // enables pure request calls
        if (typeof link === 'string') {
            var link_element = document.createElement('A');
            link_element.href = link;
            link = link_element;
        }

        if (link.tagName.toUpperCase() !== 'A') {
            throw '$.fn.pjaxr requires an anchor element';
        }

        // ignore cross origin links and pass relative IE links (cause they IE can't do code)
        if ((location.protocol !== link.protocol || location.hostname !== link.hostname) && !(link.protocol === ':' && link.hostname === '')) {
            return;
        }

        // ignore anchors on the same page
        if (link.hash && link.href.replace(link.hash, '') === location.href.replace(location.hash, '')) {
            return;
        }

        // ignore empty anchor 'foo.html#'
        if (link.href === location.href + '#') {
            return;
        }

        var url = $.isFunction(link.href) ? link.href() : link.href;

        var xhr_url = url;
        if (xhr_url.indexOf('pjaxr') == -1) {
            xhr_url += url.indexOf('?') == -1 ? '?pjaxr' : '&pjaxr';
        }

        var defaults = {
            url: xhr_url,
            plainUrl: url,
            type: 'GET',  // always GET since we currently not support other methods
            dataType: 'html'
        };

        var opts = fnPjaxr.options = $.extend(true, {}, $.ajaxSettings, defaults, $.fn.pjaxr.defaults, options);

        if (!namespace) {
            namespace = $('body').data('pjaxr-namespace') || '';
        }

        var timeoutTimer;

        opts.beforeSend = function (xhr, settings) {
            if (!fire('pjaxr:beforeSend', [xhr, settings])) {
                return false;
            }

            xhr.setRequestHeader('X-PJAX', 'true');
            xhr.setRequestHeader('X-PJAX-NAMESPACE', namespace);

            if (settings.timeout > 0) {
                timeoutTimer = setTimeout(function () {
                    if (fire('pjaxr:timeout', [xhr, opts])) {
                        xhr.abort('timeout');
                    }
                }, settings.timeout);

                // clear timeout setting so jQuery's internal timeout isn't invoked
                settings.timeout = 0;
            }

            return true;
        };

        // create pjaxr state for initial page load
        if (!fnPjaxr.state) {
            fnPjaxr.state = {
                id: uniqueId(),
                namespace: namespace,
                url: window.location.href,
                title: document.title
            };
            window.history.replaceState(fnPjaxr.state, fnPjaxr.state.title, fnPjaxr.state.url);
        }

        var xhr = fnPjaxr.xhr;

        // cancel the current running pjaxr request if there is one
        if (xhr && xhr.readyState < 4) {
            xhr.onreadystatechange = $.noop;
            xhr.abort();
        }

        // go-go-pjaxr
        xhr = fnPjaxr.xhr = $.ajax(opts);

        if (xhr.readyState > 0) {
            fire('pjaxr:start', [opts]);
            fire('pjaxr:send', [opts]);
        }

        xhr.done(function (data, textStatus, jqXHR) {
            fire('pjaxr:success', [opts]);

            var currentVersion = (typeof opts.version === 'function') ? opts.version() : opts.version;
            var latestVersion = jqXHR.getResponseHeader('X-PJAX-Version');

            // If there is a layout version mismatch, hard load the new url
            if (currentVersion && latestVersion && currentVersion !== latestVersion) {
                loadHard(opts.plainUrl);
                return;
            }

            var head_match = data.match(/<pjaxr-head>([\s\S.]*)<\/pjaxr-head>/i);
            var body_match = data.match(/<pjaxr-body>([\s\S.]*)<\/pjaxr-body>/i);

            // if response data doesn't fit, hard load the new url
            if (!head_match && !body_match) {
                loadHard(opts.plainUrl);
                return;
            }
            fire('pjaxr:success', [data, textStatus, jqXHR, opts]);

            // Clear out any focused controls before inserting new page contents.
            document.activeElement.blur();

            var stateId = uniqueId();

            if (head_match) {
                var $head = $(parseHTML(head_match[0]));
                var head_parts = processPjaxrHead('forward', $head.children(), null, null);
                var apply_head_parts = head_parts[0];
                var revert_head_parts = head_parts[1];
                var remove_head_parts = head_parts[2];
            }

            if (body_match) {
                var $body = $(parseHTML(body_match[0]));
                var body_parts = processPjaxrBody($body.children());
                var apply_body_parts = body_parts[0];
                var revert_body_parts = body_parts[1];
            }

            var namespace_match = data.match(/<pjaxr-namespace>([\s\S.]*)<\/pjaxr-namespace>/i);
            if (namespace_match) {
                namespace = $(parseHTML(namespace_match[0].replace(/(\r\n|\n|\r|\s|\t)/gm, ''))).html();
            }

            // FF bug: Won't autofocus fields that are inserted via JS.
            // This behavior is incorrect. So if there's no current focus, autofocus
            // the last field.
            //
            // http://www.w3.org/html/wg/drafts/html/master/forms.html
            $(document).find('input[autofocus], textarea[autofocus]').last().focus();

            if (typeof opts.scrollTo === 'number') {
                $(window).scrollTop(opts.scrollTo);
            }

            // enrich current state information with removal instructions
            $.extend(fnPjaxr.state, {
                head_revert: head_match ? revert_head_parts : null,
                head_remove: head_match ? remove_head_parts : null,
                body_revert: body_match ? revert_body_parts : null
            });
            if (opts.push || opts.replace) {
                window.history.replaceState(fnPjaxr.state, fnPjaxr.state.title, fnPjaxr.state.url);
            }

            fnPjaxr.state = {
                id: stateId,
                namespace: namespace,
                // by pushing the plainUrl - pjaxr param doesn't appear in the navigation bar and prevents browser caching bugs - #33
                url: opts.plainUrl,
                title: document.title,
                head_apply: head_match ? apply_head_parts : null,
                body_apply: body_match ? apply_body_parts : null
            };

            if (opts.push) {
                window.history.pushState(fnPjaxr.state, fnPjaxr.state.title, fnPjaxr.state.url);
            }
            else if (opts.replace) {
                window.history.replaceState(fnPjaxr.state, fnPjaxr.state.title, fnPjaxr.state.url);
            }

            fire('pjaxr:done', [data, textStatus, jqXHR, opts]);
        });

        xhr.fail(function (jqXHR, textStatus, errorThrown) {
            if (textStatus !== 'abort' && fire('pjaxr:fail', [jqXHR, textStatus, errorThrown, opts])) {
                loadHard(opts.plainUrl);
            }
        });

        xhr.always(function () {
            if (timeoutTimer) {
                clearTimeout(timeoutTimer);
            }

            fire('pjaxr:always', [opts]);
            fire('pjaxr:end', [opts]);
        });
    }

    function handleClick(event, options) {
        var link = event.currentTarget;

        if (link.tagName.toUpperCase() !== 'A') {
            throw '$.fn.pjaxr requires an anchor element';
        }

        // middle click, cmd click, and ctrl click should open links in a new tab as normal.
        if (event.which > 1 || event.metaKey || event.ctrlKey || event.shiftKey || event.altKey) {
            return;
        }

        // ignore prevented links
        if (event.isDefaultPrevented()) {
            return;
        }

        if (!fire('pjaxr:click')) {
            event.preventDefault();
            return;
        }

        request(link, options);

        event.preventDefault();
    }

    function processPjaxrHeadElements(elements, append) {
        var apply_head_parts = [];
        var revert_head_parts = [];
        var remove_head_parts = [];

        if (elements && elements.length > 0) {
            $.each(elements, function (index, value) {
                var $value = $(value);

                // only applied on push
                if ($value.is('title')) {
                    document.title = $value.text();
                }
                else if ($value.is('meta')) {
                    var $meta;
                    var name = $value.attr('name');
                    var property = $value.attr('property');

                    if (name) {
                        $meta = $('head > meta[name="' + name + '"]');
                    }
                    else if (property) {
                        $meta = $('head > meta[property="' + property + '"]');
                    }

                    if ($meta !== undefined) {
                        if ($meta.length > 0) {
                            remove_head_parts.push(outerHTML($meta));
                            $meta.remove();
                        }
                        else {
                            revert_head_parts.push(outerHTML($value));
                        }

                        if (append === true) {
                            $('head').append($value);
                            apply_head_parts.push(outerHTML($value));
                        }
                    }
                }
                else if ($value.is('link')) {
                    var link_href = $value.attr('href');
                    if (link_href) {
                        var $link = $('head > link[href="' + link_href + '"]');

                        if ($link.length > 0) {
                            remove_head_parts.push(outerHTML($link));
                            $link.remove();
                        }
                        else {
                            revert_head_parts.push(outerHTML($value));
                        }

                        if (append === true) {
                            $('head').append($value);
                            apply_head_parts.push(outerHTML($value));
                        }
                    }
                }
                else if ($value.is('script')) {
                    var script_src = $value.attr('src');
                    if (script_src) {
                        var $script = $('head > script[src="' + script_src + '"]');

                        if ($script.length > 0) {
                            remove_head_parts.push(outerHTML($script));
                            $script.remove();
                        }
                        else {
                            revert_head_parts.push(outerHTML($value));
                        }

                        if (append === true) {
                            $('head').append($value);
                            apply_head_parts.push(outerHTML($value));
                        }
                    }
                }
                else if ($value.is('style')) {
                    if (append === true) {
                        $('head').append($value);
                        apply_head_parts.push(outerHTML($value));
                        revert_head_parts.push(outerHTML($value));
                    }
                }
            });
        }

        return [apply_head_parts, revert_head_parts, remove_head_parts];
    }

    function processPjaxrHead(direction, apply_elements, revert_elements, remove_elements) {
        var apply_head_parts = [];
        var revert_head_parts = [];
        var remove_head_parts = [];

        // cleanup head elements
        if (direction === 'forward') {
            $('head > [data-remove-on-pjaxr]').each(function () {
                var $this = $(this);
                remove_head_parts.push(outerHTML($this));
                $this.remove();
            });
        }
        else if (direction === 'back') {
            var revert_result = processPjaxrHeadElements(revert_elements, false);
            var remove_result = processPjaxrHeadElements(remove_elements, true);

            // there are no apply elements on back processing
            $.extend(revert_head_parts, revert_result[1]);
            $.extend(remove_head_parts, remove_result[2]);
        }

        var apply_result = processPjaxrHeadElements(apply_elements, true);
        $.extend(apply_head_parts, apply_result[0]);
        $.extend(revert_head_parts, apply_result[1]);
        $.extend(remove_head_parts, apply_result[2]);

        return [apply_head_parts, revert_head_parts, remove_head_parts];
    }

    function processPjaxrBody(elements) {
        var apply_body_parts = [];
        var revert_body_parts = [];

        if (elements && elements.length > 0) {
            $.each(elements, function (index, value) {
                var $value = $(value);
                var id = $value.attr('id');
                if (id) {
                    var $target = $('#' + id);
                    if ($target.length > 0) {
                        revert_body_parts.push(outerHTML($target));
                        try {
                            $target.html($value.html());
                        }
                        catch (error) {
                            if (window.console) {
                                console.error(error);
                            }
                        }
                        apply_body_parts.push(outerHTML($target));
                    }
                }
            });
        }

        return [apply_body_parts, revert_body_parts];
    }

    // helper to trigger jQuery events
    function fire(type, args) {
        var event = $.Event(type);
        $(document).trigger(event, args);
        return !event.isDefaultPrevented();
    }

    // wrapper to save some arguments
    function parseHTML(html) {
        return $.parseHTML(html, document, true)
    }

    // helper to get the outer html
    function outerHTML(element) {
        return element.clone().wrap('<p>').parent().html();
    }

    // generate a unique id for state object based on the current timestamp
    function uniqueId() {
        return (new Date()).getTime();
    }

    // hard load to new state without pjaxr
    // replace would brick expected history behavior - see: #17
    function loadHard(url) {
        window.location.href = url;
    }

    // takes care of the back and forward functionality
    function onPjaxrPopstate(event) {
        var state = event.state;

        if (state) {
            // when coming forward from a separate history session, will get an
            // initial pop with a state we are already at. Skip reloading the current page.
            if (initialPop && initialURL == state.url) {
                return;
            }

            fire('pjaxr:start', [fnPjaxr.options]);

            // title is always set, no check
            document.title = state.title;

            // determine whether to go back or forth
            var direction = fnPjaxr.state.id < state.id ? 'forward' : 'back';

            // null check inside
            processPjaxrHead(direction, state.head_apply, state.head_revert, state.head_remove);

            var body = direction === 'forward' ? state.body_apply : state.body_revert;
            if (body && body.length > 0) {
                processPjaxrBody(body);
            }

            fnPjaxr.state = state;
            namespace = state.namespace;

            // force reflow / relayout before the browser tries to restore the scroll position.
            document.body.offsetHeight;

            fire('pjaxr:end', [fnPjaxr.options]);
        }
        initialPop = false;
    }

    // helper to extract pjaxr version from head meta tag
    function findVersion() {
        return $('meta').filter(function () {
            return String($(this).attr('http-equiv')).toUpperCase() === 'X-PJAX-VERSION';
        }).attr('content');
    }

    var initialPop = true;
    var initialURL = window.location.href;
    var initialState = window.history.state;
    var namespace;

    // initialize $.fnPjaxr.state if possible
    // happens when reloading a page and coming forward from a different session history.
    if (initialState) {
        fnPjaxr.state = initialState;
    }

    // non-webkit browsers don't fire an initial popstate event
    if ('state' in window.history) {
        initialPop = false;
    }

    // add the state property to jQuery's event object so we can use it in $(window).on('popstate', ...)
    if ($.inArray('state', $.event.props) < 0) {
        $.event.props.push('state');
    }

    // enables pushState behavior
    function enable() {
        $.fn.pjaxr = fnPjaxr;
        $.fn.pjaxrReady = fnPjaxrReady;
        $.fn.pjaxrAlways = fnPjaxrAlways;
        $.fn.pjaxr.click = handleClick;
        $.fn.pjaxr.request = request;
        $.fn.pjaxr.enable = $.noop;
        $.fn.pjaxr.disable = disable;
        $.fn.pjaxr.defaults = {
            timeout: 650,
            push: true,
            replace: false,
            scrollTo: 0,
            version: findVersion
        };
        $(window).on('popstate.pjaxr', onPjaxrPopstate);
    }

    // disable pushState behavior
    function disable() {
        $.fn.pjaxr = function () {
            return this;
        };
        $.fn.pjaxrReady = function (func) {
            return $(document).ready(func);
        };
        $.fn.pjaxrAlways = function (func) {
            return $(document).ready(func);
        };
        $.fn.pjaxr.enable = enable;
        $.fn.pjaxr.disable = $.noop;
        $(window).off('popstate.pjaxr', onPjaxrPopstate);
    }

    // is pjaxr supported by this browser?
    $.support.pjaxr = window.history && window.history.pushState && window.history.replaceState &&
        // pushState isn't reliable on iOS until 5
        !navigator.userAgent.match(/((iPod|iPhone|iPad).+\bOS\s+[1-4]|WebApps\/.+CFNetwork)/);
    // initial executes enable / disable pjaxr when the script gets loaded
    $.support.pjaxr ? enable() : disable();
})(jQuery);
